spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java

```diff
@@ -905,9 +911,22 @@ protected void verifyResultCollectionConsistsOfMessages(Collection<?> elements)
 				"The expected collection of Messages contains non-Message element: " + commonElementType);
 	}
 
-	protected Long obtainGroupTimeout(MessageGroup group) {
-		return this.groupTimeoutExpression != null
-				? this.groupTimeoutExpression.getValue(this.evaluationContext, group, Long.class) : null;
+	protected Object obtainGroupTimeout(MessageGroup group) {
+		if (this.groupTimeoutExpression != null) {
+			Object timeout = this.groupTimeoutExpression.getValue(this.evaluationContext, group);
+			if (timeout instanceof Date) {
+				return timeout;
+			}
+			else if (timeout != null) {
+				try {
+					return Long.parseLong(timeout.toString());
+				}
+				catch (NumberFormatException ex) {
+					throw new IllegalStateException("Error evaluating 'groupTimeoutExpression'", ex);
```

Comment:
Should we consider `Duration` as well?

Reply:
No, that one goes already to a `fixedDelay/fixedRapte` scheduling variants, not a single one as our logic here is based on.